src/main/kotlin/com/terraformation/backend/nursery/db/BatchStore.kt

```diff
@@ -148,16 +161,104 @@ class BatchStore(
     )
   }
 
-  private fun insertQuantityHistoryRow(row: BatchesRow, historyType: BatchQuantityHistoryType) {
+  fun updateQuantities(
+      batchId: BatchId,
+      version: Int,
+      germinating: Int,
+      notReady: Int,
+      ready: Int,
+      historyType: BatchQuantityHistoryType,
+  ) {
+    if (germinating < 0 || notReady < 0 || ready < 0) {
+      throw IllegalArgumentException("Quantities may not be negative")
+    }
+
+    requirePermissions { updateBatch(batchId) }
+
+    dslContext.transaction { _ ->
+      updateVersionedBatch(batchId, version) { update ->
+        update
+            .set(GERMINATING_QUANTITY, germinating)
+            .set(NOT_READY_QUANTITY, notReady)
+            .set(READY_QUANTITY, ready)
+            .let {
+              if (historyType == BatchQuantityHistoryType.Observed) {
+                it.set(LATEST_OBSERVED_GERMINATING_QUANTITY, germinating)
+                    .set(LATEST_OBSERVED_NOT_READY_QUANTITY, notReady)
+                    .set(LATEST_OBSERVED_READY_QUANTITY, ready)
+                    .set(LATEST_OBSERVED_TIME, clock.instant())
+              } else {
+                it
+              }
+            }
+      }
+
+      insertQuantityHistoryRow(batchId, germinating, notReady, ready, historyType)
+    }
+  }
+
+  /**
+   * Applies an update to a batch if the caller-supplied version number is up to date.
+   *
+   * Automatically updates `modified_by`, `modified_time`, and `version`.
+   *
+   * @param func Function to add `set` statements to an `UPDATE` query. This is called with the
+   * [BATCHES] table as its receiver so that lambda functions can refer to column names without
+   * having to qualify them (that is, `set(FOO, value)` instead of `set(BATCHES.FOO, value)`).
+   * @throws BatchStaleException The batch's version number in the database wasn't the same as
+   * [version].
+   */
+  private fun updateVersionedBatch(
+      batchId: BatchId,
+      version: Int,
+      func: Batches.(UpdateSetFirstStep<BatchesRecord>) -> UpdateSetMoreStep<BatchesRecord>
+  ) {
+    val rowsUpdated =
+        dslContext
```

Comment:
do we need to lock the row or will concurrency not be an issue.

Reply:
The database has implicit locking that takes care of this. Basically if one transaction updates a row and another transaction is doing a write whose success or failure depends on one of the values updated by the first transaction, the second write will block until the first transaction either commits or rolls back.

You can play with this in `psql` by connecting to thee same database in two windows and doing something like

```
(window 1)
create table foo (col integer);
insert into foo values (1);
begin;
update foo set col = 2 where col = 1;

(window 2)
update foo set col = 2 where col = 1;
-- this blocks even though you never asked for a lock anywhere

(window 1)
commit;

(window 2 unblocks and shows 0 rows updated)
```

Reply:
Thanks so much, figured something like this is going on.
From a write/read concurrency perspective - I guess it's ok for the read to not wait for the write (with the caveat that the read could potentially be stale .. but avoids slowing down all reads).
I'm overthinking all this for sure.

Reply:
https://www.postgresql.org/docs/13/transaction-iso.html has a more detailed discussion of this stuff if you want to _really_ overthink it.

And it's fine if reads return stale data. No different from if the read came in a few milliseconds earlier and didn't happen to overlap with the write.

Reply:
Oh, and the reason I did it this way rather than explicit locking is because of the "optimistic" part of the equation: 99.9% of the time, there won't be another user editing the same batch and this will just succeed without anything getting blocked. Grabbing an explicit lock means paying the performance penalty on every write (at minimum, an extra database round-trip).

Reply:
Makes total sense, thanks for the clarification and the document links @sgrimm